### PR TITLE
Fix default dtype in shape analysis

### PR DESCRIPTION
**Pull request by eellison**
Fix for https://github.com/pytorch/pytorch/issues/18823

Previously we were setting the dtype to Float when in torchscript the default is double. When the problem in https://github.com/pytorch/pytorch/issues/17662 gets landed, we will have to reevalute (and this test will fail).  

We should still be consistent in shape_analysis in the meantime.

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -8065,6 +8065,18 @@ def test_rand():
             return a + 1.0 - a
 
         self.checkScript(test_rand, ())
+        fn = torch.jit.script(test_rand)
```
Comment (suo) [inline test/test_jit.py]: Can you add a comment for these tests explaining what they are intended to test?

**Review by suo — APPROVED** [review]

**Review by apaszke — COMMENTED** [review]
This is wrong... You saw the test failure because we set the default type to `double` inside our tests (for gradchecks), while it is usually `float` in user programs. This is controlled by `torch.set_default_dtype`, so we shouldn't rely on that value being statically available...

**Comment by eellison** [comment]
Ah word, I did not know that. Shape analysis is only run at runtime so I can change the dtype to check what the current default dtype is.  But really, we should fix 17662

**Comment by apaszke** [comment]
No, you can't, unless we make this part of `ArgumentSpec`. It's not like you can set this once at the beginning of the script (although that's the usual use case). It can really change at any point, including after you generate a specialized graph.

**Comment by eellison** [comment]
It seems extremely strange to me to change the default dtype during runtime

**Comment by eellison** [comment]
> It seems extremely strange to me to change the default dtype during training / inference.

**Pull request closed without merging**